### ISSUE TITLE
LineagePulse

### ISSUE THREAD

**Issue by davidsebfischer** [Bioconductor/Contributions#521]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/YosefLab/LineagePulse

Confirm the following by editing each check box to '[x]'

- [x ] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [ x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x ] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [ x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [ x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [ x] The 'devel' branch for new packages and features.
- [ x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x ] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by davidsebfischer** [comment]
Do I have to take any further steps here or should I just wait? Thanks, David.

**Comment by nturaga** [comment]
Hi @davidsebfischer I'll post your package review today.

**Comment by nturaga** [comment]
# LineagePulse Review

* R CMD build : ok

* R CMD check: 

    Consider adding
        importFrom("graphics", "contour")
        importFrom("stats", "density")

* R CMD INSTALL: ok

## DESCRIPTION

* ok.

* Don't need Author + Maintainer fields if you are using Authors@R.

## NAMESPACE

* Ok

## man

* Good.

## vignette

* Vignette has a very high level explaination of what is happening. Not enough detail.

* Make sure it's formatted correctly, there is a "true" on top of the page.

## R 

* Maybe worth looking at the class `SingleCellExperiment` https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html.

* Use `message` instead of `print` statements when you want to show text on the stdout, eg: line 38 , main_LineagePulse.R 

    ```
    if(boolVerbose) print(strMessage)
    ```

    Do the same thing for all other print messages as well. Try this to locate all your print statements,
    `grep -r 'print' R/*`.

* `dfAnnotation` should go in the colData of your class `SummariedExperiment`, and should be extracted from there as need in the runLineagePulse function.

* Use `is(counts, "SummarizedExperiment")` in l261 of main_LineagePulse.R.

* Add "accessor" function to the slots in your object `objLP` so that you are able to access the slots with a call `strReport(objLP)` instead of `objLP@strReport`. Use the accessor you wrote `get_strReport`.

* Add and use "setters" as needed, e.g.,

  ```
  `set_strReport<-` <- function(objLP, value) {
      objLP@strReport <- value
      objLP
  }
  ```

  Note that the `set_` and `get_` prefix is redundant in R

  ```
  strReport <- function(objLP, value)     ## .. getter
  `strReport<-` <- function(objLP, value) ## .. setter
  strReport(objLP) <- strReport(objLP)    ## no-op, set strReport of objLP to current value
  ```
      

* l80, main_LineagePulse.R, this is done automatically in R, and there is no need to remove objects or garbage collect explicitly. These arguments are being passed into the function, so the calls to `rm` don't do anything.

    ```
    # Clear memory
    rm(counts)
    rm(matPiConstPredictors)
    rm(dfAnnotation)
    rm(lsProcessedSCData)
    gc()
    ```

    Do this in every place you use `rm` to remove an object.


* Use `seq_len` instead of (srcLineagePulse_calcPostDrop.R, l34)

    ```
    scaNumSamples <- length(vecMu)
    
    # Compute probability of zero counts under 
    # negative binomial model.
    vecNBZero <- (vecDisp/(vecDisp+vecMu))^vecDisp
    # Compute posterior of drop-out.
    vecZ <- sapply(seq(1,scaNumSamples), function(j){
    ```

    this can be written as

    ```
    vecNBZero <- (vecDisp/(vecDisp+vecMu))^vecDisp
    vecZ <- sapply(seq_len(vecMu), function(j) {
        ...
    ```

*   This entire function can be vectorized and the `sapply` can be probably avoided.

    ```
    scaZ <- vecDrop/(vecDrop + (1-vecDrop)*vecNBZero)
    vecZ <- rep(NA_real_, length(vecMu))
    vecZ[ vecboolNotZero ] <- 0
    vecZ[ vecboolZero ] <- scaZ[ vecboolZero ]
    ```

    This is your implementation

    ```	    
    vecZ <- sapply(seq(1,scaNumSamples), function(j){
        if(vecboolNotZero[j]){
            scaZ <- 0
        } else if(vecboolZero[j]) {
            scaZ <- sum(vecDrop[j]/(vecDrop[j] + (1-vecDrop[j])*vecNBZero[j])) *
                1/length(j)
        } else {
            scaZ <- NA
        }
        return(scaZ)
    })
    ```


* You can replace this with `ncol`, and `nrow`.

    ```
    scaNumGenes <- dim(matCounts)[1]
    scaNumCells <- dim(matCounts)[2]
    ```

* In srcLineagePulse_initializeImpulseParameters.R, line 60, `seq(1, scaGroups)` can be `seq_len(scaGroups)`. Do this every place. Look into using `seq_along` and `seq_len` where needed. Try not to use `1:x`.

* `stop` and `warning` do not need `paste0` inside it. It works without it. srcLineagePulse_decompressParameters.R, l215. Do this everywhere as needed. Do not use the function `print` to pass a message to the stdout. Instead use `message`, everywhere.


* In srcLineagePulse_evalLogLikZINB.R, l228, the function again calls many lapply functions within the bplapply call, with no documentation, making it very hard to follow what is happening.


* Another vectorization is in srcLineagePulse_evalImpulseModel.R


 ```
 evalImpulseModel <- function(vecImpulseParam,
			      vecTimepoints){

     t1 <- (vecTimepoints - vecImpulseParam[6])
     t2 <- (vecTimepoints - vecImpulseParam[7])

     vecImpulseValue <- 
	 (1/vecImpulseParam[4]) * 
	     (vecImpulseParam[3] + (vecImpulseParam[4]-vecImpulseParam[3])*
		  (1/(1+exp(-vecImpulseParam[1]*t1)))) *
	     (vecImpulseParam[5] + (vecImpulseParam[4]-vecImpulseParam[5])*
		  (1/(1+exp(vecImpulseParam[2]*t2))))

     return(vecImpulseValue)
 }
 ```

Your implementation was,

 ```
 evalImpulseModel <- function(vecImpulseParam,
			      vecTimepoints){

     vecImpulseValue <- sapply(vecTimepoints, function(t){
	 (1/vecImpulseParam[4]) * 
	     (vecImpulseParam[3] + (vecImpulseParam[4]-vecImpulseParam[3])*
		  (1/(1+exp(-vecImpulseParam[1]*(t-vecImpulseParam[6]))))) *
	     (vecImpulseParam[5] + (vecImpulseParam[4]-vecImpulseParam[5])*
		  (1/(1+exp(vecImpulseParam[2]*(t-vecImpulseParam[7])))))
     })

     return(vecImpulseValue)
 }
 ```
  
Consider vectorizing as much as possible in your entire codebase.

**Comment by nturaga** [comment]
@davidsebfischer When you are done making the required changes on your package. Please be sure to check the build report, and provide post a summary of all the changes you have made. I will review your package after receiving the summary from you.

**Comment by davidsebfischer** [comment]
@nturaga Here the summary of how I addressed all suggestions:

1. LineagePulse Review
    importFrom("graphics", "contour") - This suggestion was thrown because a data frame column was labelled contour and called in ggplot2. I renamed that column.
    importFrom("stats", "density") - Added this.

2. DESCRIPTION
    Don't need Author + Maintainer fields if you are using Authors@R. - removed

3. vignette
    Vignette has a very high level explaination of what is happening. Not enough detail.
    Make sure it's formatted correctly, there is a "true" on top of the page.
- added more example usage sections in vignette. Please let me know if you have something different in mind.

4. R
    # Maybe worth looking at the class SingleCellExperiment https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html. - supported as input

    # Use message instead of print statements when you want to show text on the stdout - done

    # dfAnnotation should go in the colData of your class SummariedExperiment, and should be extracted from there as need in the runLineagePulse function. - I have a separate class LineagePulse which is tailored to the fitting problem encountered here. I therefore keep dfAnnotation as a separate object. Is it necessary to change this?

    # Use is(counts, "SummarizedExperiment") in l261 of main_LineagePulse.R. - done

    # Add "accessor" function to the slots in your object objLP so that you are able to access the slots with a call strReport(objLP) instead of objLP@strReport. Use the accessor you wrote get_strReport. - done

    # Add and use "setters" as needed, e.g., - done for througout entire package for all set operations in all functions.

    # l80, main_LineagePulse.R, this is done automatically in R, and there is no need to remove objects or garbage collect explicitly. These arguments are being passed into the function, so the calls to rm don't do anything. - removed

    # Use seq_len instead of (srcLineagePulse_calcPostDrop.R, l34) - used seq_len and seq_along throughout code.
    
   # This entire function can be vectorized and the sapply can be probably avoided. - replaced

    # You can replace dim(matCounts)[x] with ncol, and nrow. - done throughout code

    # In srcLineagePulse_initializeImpulseParameters.R, line 60, seq(1, scaGroups) can be seq_len(scaGroups). Do this every place. Look into using seq_along and seq_len where needed. Try not to use 1:x. - used seq_len and seq_along throughout code.

    # stop and warning do not need paste0 inside it. It works without it. - removed paste0 throughout

    # In srcLineagePulse_evalLogLikZINB.R, l228, the function again calls many lapply functions within the bplapply call, with no documentation, making it very hard to follow what is happening. - wrote two wrappers for lapply calls in this section.

    # Another vectorization is in srcLineagePulse_evalImpulseModel.R - replaced

**Comment by davidsebfischer** [comment]
@nturaga Have you had a chance to look at this? David

**Comment by nturaga** [comment]
Hi @davidsebfischer I've done an initial first review of the replaced code and it looks good. 

Updates from first review:

* Your vigenette still has `print(` and `paste0`. You only need `print` to concatenate strings.

**Comment by davidsebfischer** [comment]
Thanks for your input, @nturaga, but I cannot find a way to use print() on vectors of strings without a call to paste0(): 
print("a", "b") yields an error and print(c("a", "b")) an undesired result.

**Comment by nturaga** [comment]
Yes, print doesn't work that way. But `cat` does.

```
 > cat("Mean fitted mean parameter: ", round(mean(matMeanParamFit),1), "\n")
 Mean fitted mean parameter:  307.4
> cat("Minimum fitted mean parameter: ", round(min(matMeanParamFit),1), "\n")
 Minimum fitted mean parameter:  77.6
```

**Comment by davidsebfischer** [comment]
Thanks @nturaga , I replaced the print(paste0(...)) calls in the vignette by cat(...) calls.

**Comment by nturaga** [comment]
@davidsebfischer Remove the "html" document from your vignette's folder. It is not needed. Just the Rmd is enough.

You package looks in good shape for acceptance. Issue a version bump please before we proceed.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/davidsebfischer.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://docs.google.com/forms/d/e/1FAIpQLSdlTbNjsQJDp0BA480vo4tNufs0ziNyNmexegNZgNieIovbAA/viewform

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"LineagePulse\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/LineagePulse

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.
